### ISSUE TITLE
Ladders need to exit on top

### ISSUE THREAD

**Issue by D3nnis3n** [StrangeLoopGames/EcoIssues#9908]
Watch the video: https://drive.google.com/file/d/1_dFw4uTE7onEwokOkdO2FdllXG9pOWxg/view?usp=sharing

Why do i not get onto the top dirt block automatically when i am on top of the ladder but need to jump and fail several times doing so? When trying to go down why can i not walk near the ladder and be on it to climb down, but can't get onto the ladder from top at all? I know, no fall damage, but please - realism XD

**Comment by elfl0rd** [comment]
Yes, and another problem, leaving in situations like this
![unknown](https://user-images.githubusercontent.com/4980243/47746728-bc2c6a80-dc97-11e8-9029-8bd7ae5247ba.png)
I want direct attach w\o pressing buttons, you can attach to it when fall\jump

**Comment by elfl0rd** [comment]
I prefer ladders in Rust, really usefull.

**Comment by D3nnis3n** [comment]
Don't know Rust, but most games just put you on the ladder working towards it and eject you from the ladder when at top or bottom - should be the thing eco should do, too.

**Comment by johnkslg** [comment]
@MazonDel can you post an analysis of how you want to tackle this one before starting and Ill review?

**Comment by MazonDel** [comment]
@johnkslg Regarding an analysis for this:

In the PlayerMotorBehaviorKinematic.TryLadderMovement() we already successfully identify when the player has reached the top of the ladder that currently exists. The below analysis achieves the goal of leaving the ladder without modifying anything about the ladders themselves.

What seems like it would best replicate the behavior at the bottom of the ladder (where the player immediately steps off and moves) without causing incidents where the players step off the ladder when they might want to, would be to perform a directionality check once we've reached the top. If the player is still looking directly at the ladder (+/- some angle) then nothing happens, they just stop moving like currently. But if they are looking in any other direction, then they cleanly come off the ladder with some velocity in the direction they are facing. The exception to this would be if they are pressing D, which currently causes the player to descend the ladder.

If we want to consider adding trigger colliders to the ladders, this opens the possibility for smoothly latching onto the ladder in addition to the smooth disconnect at the top.

Alternative to trigger colliders, every time CalculateMovement() or TryLadderMovement() is executed, a check of the block immediately in front of the player can be performed, if it is a ladder then a more precise directionality check performed to determine if the player is looking AT the ladder instead of past it. If the player is looking at the ladder and they enter within a certain very small range, it can be assumed they want to mount the ladder. We can also do the directionality check described earlier at any point along the ladder height to use as a trigger to dismount the ladder, but I'd caution against this expansion and just leave it to using jump to dismount at a non-top height.

Of note as well, we may or may not want to consider the possibility of altering the jump system such that when jumping "off" a ladder, the player has more horizontal velocity in the direction they are facing than in the vertical direction. This could make it easier to dismount the ladder at specific locations that are neither top nor bottom.

**Comment by johnkslg** [comment]
This makes sense, if youre looking away from the ladder it can detach you from it.   If you get to the top and youre still pressing up, and there's enough empty space above you, it could start a sequence where it animates you to the position atop the ladder. This is for the case where the ladder is up the side of a wall, and you want them to get atop the wall when they reach the top of the ladder.

**Comment by johnkslg** [comment]
@D3nnis3n video seems to be missing now

**Comment by SlayksWood** [comment]
close this one, we have another issue to improve it #18485, #18501, #18502
